### PR TITLE
Week 3 assignments part 1

### PR DESCRIPTION
**Pull request by jepetko**
- 1 Lesson: Growing Complexity Guided by Tests
- 2 Lesson: Interactive Debugging for Solution Discovery
- 3 Lesson: Respond to Feature Changes
- 4 Lesson: Transactions
- 5 Assignment: User can reorder videos in the queue
- 6 Lesson: Structural Refactor
- 7 Lesson: Skinny Controller, Fat Model
- 8 Lesson: Fat Controllers in the Wild
- 9 Assignment: Push down complexity
- 10    Assignment: User can change video ratings in the queue
- 11    Code Review

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -3,8 +3,51 @@ class User < ActiveRecord::Base
   # validates password and password_confirmation; no validates_presence_of necessary!!!
   has_secure_password validations: true
   has_many :queue_items, -> { order('order_value') }
+  has_many :reviews
 
   validates_presence_of :email, :full_name
   validates_uniqueness_of :email
 
+  def update_queue_items(queue_items_hash)
```
Comment (tomtomecek) [inline app/models/user.rb]: Pushing this down to the model layer is actually considered as bad practice. I am referring to [this assignment](https://www.gotealeaf.com/lessons/00bcc3b5/assignments/1569) especially solution video from 0:17 to 1:07.

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -22,4 +22,8 @@ def select_collection_rating_values
       [ pluralize(rating, 'star'), rating]
     end
   end
+
+  def build_queue_items_json(queue_items)
```
Comment (tomtomecek) [inline app/helpers/application_helper.rb]: I don't think this is needed.

File: app/models/user.rb [hunk 3]
```diff
@@ -3,8 +3,51 @@ class User < ActiveRecord::Base
...
+    ActiveRecord::Base.transaction do
+      queue_items_hash.each do |queue_item_element|
+        queue_item = queue_items.find_by(id: queue_item_element['id'])
+        next if queue_item.nil?
+        update_order_value queue_item, queue_item_element['order_value']
```
Comment (tomtomecek) [inline app/models/user.rb]: `update_order_value` + `update_rating`= a lot of complex/complicated code that can be actually refactored by using mass assignment and `update!`. Check the solution video and see if you like it better.

File: app/models/user.rb [hunk 4]
```diff
@@ -3,8 +3,51 @@ class User < ActiveRecord::Base
...
+    end
+    queue_item.update_attributes! order_value: order_value if !queue_item.nil?
+  end
+
+  def update_rating(queue_item, rating)
+    if !rating.nil?
+      rating = rating.to_i
+      video_reviews = reviews.where(video: queue_item.video)
+      if video_reviews.first
+        video_reviews.first.update_attributes! rating: rating
```
Comment (tomtomecek) [inline app/models/user.rb]: `.update_attributes!` can be just `.update!`. This is valid replacement in Rails 4.

File: app/models/user.rb [hunk 5]
```diff
@@ -3,8 +3,51 @@ class User < ActiveRecord::Base
...
+  end
+
+  private
+
+  def update_order_value(queue_item, order_value)
+    order_value = order_value.to_i
+    if order_value == 0
+      raise ArgumentError.new 'invalid order_value'
+    end
+    queue_item.update_attributes! order_value: order_value if !queue_item.nil?
```
Comment (tomtomecek) [inline app/models/user.rb]: `if !` is just `unless` in Ruby. But anyway `if !queue_item.nil?` you could just do `if queue_item` - You should get used that in Ruby, everything is truthy except nil and false. So if queue_item is nil then its not going to follow `if` to truthy statement.

File: app/views/queue_items/index.html.haml [hunk 6]
```diff
@@ -18,16 +18,17 @@
...
-                      = f.select :rating, select_collection_rating_values, selected: queued_video.video_rating, hide_label: true
+                      = f.select 'queue_items[][rating]', select_collection_rating_values, selected: queued_video.video_rating, hide_label: true
                     %td
-                      = link_to queued_video.video.category.name, category_path(queued_video.video.category)
+                      = link_to queued_video.video.category.name, category_path(queued_video.video.category) unless queued_video.video.category.nil?
                     %td
                       = link_to queued_video, method: :delete do
                         %i.glyphicon.glyphicon-remove
-            = button_to "Update Instant Queue", nil, class: "btn btn-default"
+            = f.button 'Update Instant Queue', class: 'btn btn-default'
```
Comment (tomtomecek) [inline app/views/queue_items/index.html.haml]: Maybe `f.submit` would be even better :)

File: spec/controllers/queue_items_controller_spec.rb [hunk 7]
```diff
@@ -159,4 +159,128 @@
...
+  describe 'PUT :update' do
+
+    let(:user) { user = Fabricate(:user) }
+    let(:queue_items) { queue_items = user.queue_items }
+    let(:first) { first = queue_items.first }
+    let(:sec) { sec = queue_items[1] }
+    let(:third) { third = queue_items.last }
+    before do
+      3.times do |i|
+        Fabricate(:queue_item, user: user, order_value: i+1)
```
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: There should be spaces around operators.

File: spec/controllers/queue_items_controller_spec.rb [hunk 8]
```diff
@@ -159,4 +159,128 @@
     end
   end
 
+  describe 'PUT :update' do
+
+    let(:user) { user = Fabricate(:user) }
+    let(:queue_items) { queue_items = user.queue_items }
+    let(:first) { first = queue_items.first }
```
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: You can just do `let(:first) { queue_items.first }` - no need to assign that to a local variable `first`.

This comment is for all these `let`'s.

File: spec/controllers/queue_items_controller_spec.rb [hunk 9]
```diff
@@ -159,4 +159,128 @@
...
+    let(:sec) { sec = queue_items[1] }
```
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: In Rails, you can even do `queue_items.second`,`.third` etc..

File: spec/controllers/queue_items_controller_spec.rb (hunk 7 above)
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: Also, You can use advantage of Fabrication sequences - [check this out](http://www.fabricationgem.org/#!sequences) - at the bottom especially. And then you could just do `Fabricate.times(3, :queue_item, user: user)`. I think it should work, haven't tried it yet.

File: spec/controllers/queue_items_controller_spec.rb [hunk 10]
```diff
@@ -159,4 +159,128 @@
...
+    end
+
+    context 'for authorized user' do
+      before do
+        login_user user
+      end
+      context 'for valid order values' do
+        it 'updates the order of the queue items' do
+          post :update_queue, queue_items: [  {id: first.id, order_value: 1},
+                                        {id: sec.id, order_value: 3},
```
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: This and next spec should have these additional queue_items params aligned to the first - like you have on lines 198 - 201

File: spec/controllers/queue_items_controller_spec.rb [hunk 11]
```diff
@@ -159,4 +159,128 @@
...
+    let(:sec) { sec = queue_items[1] }
+    let(:third) { third = queue_items.last }
+    before do
+      3.times do |i|
+        Fabricate(:queue_item, user: user, order_value: i+1)
+      end
+    end
+
+    context 'for authorized user' do
+      before do
```
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: This could easily be a 1 liner before block.

File: spec/models/review_spec.rb [hunk 12]
```diff
@@ -8,11 +8,14 @@
   it { should respond_to :rating }
   it { should validate_presence_of :content }
   it { should validate_presence_of :rating }
+  it { should validate_presence_of :video }
 
-  context 'for rating values between 0 and 5' do
+  let(:video) { video = Fabricate(:video) }
+
+  context 'for rating values between 1 and 5' do
```
Comment (tomtomecek) [inline spec/models/review_spec.rb:16]: You could use at your advantage shoulda matchers - and validate inclusion of. Check out should matchers docs for this and also Rails validations guides to see validation on inclusion. Basically you select an array of options that can be only valid for the column :) hope this helps you.

File: app/models/user.rb (hunk 1 above)
Comment (jepetko) [inline app/models/user.rb]: fixed

File: app/helpers/application_helper.rb (hunk 2 above)
Comment (jepetko) [inline app/helpers/application_helper.rb]: fixed

File: app/models/user.rb (hunk 3 above)
Comment (jepetko) [inline app/models/user.rb]: fixed

File: app/models/user.rb (hunk 4 above)
Comment (jepetko) [inline app/models/user.rb]: fixed

File: app/models/user.rb (hunk 5 above)
Comment (jepetko) [inline app/models/user.rb]: fixed

File: app/views/queue_items/index.html.haml (hunk 6 above)
Comment (jepetko) [inline app/views/queue_items/index.html.haml]: fixed

File: spec/controllers/queue_items_controller_spec.rb (hunk 10 above)
Comment (jepetko) [inline spec/controllers/queue_items_controller_spec.rb]: fixed

File: spec/controllers/queue_items_controller_spec.rb (hunk 8 above)
Comment (jepetko) [inline spec/controllers/queue_items_controller_spec.rb]: fixed

File: spec/controllers/queue_items_controller_spec.rb (hunk 9 above)
Comment (jepetko) [inline spec/controllers/queue_items_controller_spec.rb]: fixed

File: spec/controllers/queue_items_controller_spec.rb (hunk 11 above)
Comment (jepetko) [inline spec/controllers/queue_items_controller_spec.rb]: ?

Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: `before { login_user user }`

File: spec/controllers/queue_items_controller_spec.rb (hunk 7 above)
Comment (jepetko) [inline spec/controllers/queue_items_controller_spec.rb]: it doesnt work. fabricator:

``` ruby
Fabricator(:queue_item) do
  video
  user
  order_value Fabricate.sequence(:number, 1)
end
```

File: spec/models/review_spec.rb (hunk 12 above)
Comment (jepetko) [inline spec/models/review_spec.rb:16]:
``` ruby
  it { should validate_inclusion_of(:rating).in_array((1..5).to_a) }
```

not working. Method validate_inclusion_of undefined. Error:

NoMethodError: undefined method `validate_inclusion_of' for #<RSpec::Core::ExampleGroup::Nested_8:0x00000007c0a588>
./spec/models/review_spec.rb:12:in`block (2 levels) in <top (required)>'

File: spec/controllers/queue_items_controller_spec.rb (hunk 7 above)
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: Try this:

``` ruby
Fabricate(:queue_item) do
  video
  user
  order_value { sequence(:order_value) { |i| i + 1 } }
end
```

File: spec/models/review_spec.rb (hunk 12 above)
Comment (tomtomecek) [inline spec/models/review_spec.rb:16]: https://github.com/thoughtbot/shoulda-matchers/blob/master/lib/shoulda/matchers/active_model/validate_inclusion_of_matcher.rb#L22

also you are using almost 1 year old version of shoulda-matchers might be that.

File: spec/controllers/queue_items_controller_spec.rb (hunk 7 above)
Comment (jepetko) [inline spec/controllers/queue_items_controller_spec.rb]: fixed

File: spec/controllers/queue_items_controller_spec.rb (hunk 11 above)
Comment (jepetko) [inline spec/controllers/queue_items_controller_spec.rb]: fixed

File: spec/models/review_spec.rb (hunk 12 above)
Comment (jepetko) [inline spec/models/review_spec.rb:16]: fixed

**Pull request merged**